pkg/kubelet/cm/deviceplugin/device_plugin_handler.go

```diff
@@ -205,74 +406,73 @@ func (h *DevicePluginHandlerImpl) Allocate(pod *v1.Pod, container *v1.Container,
-	}
-	return ret, nil
+	return h.writeCheckpoint()
 }
 
-// updateAllocatedDevices updates the list of GPUs in use.
-// It gets a list of active pods and then frees any GPUs that are bound to
+// GetDeviceRunContainerOptions checks whether we have cached containerDevices
+// for the passed-in <pod, container> and returns its DeviceRunContainerOptions
+// for the found one. An empty struct is returned in case no cached state is found.
+func (h *HandlerImpl) GetDeviceRunContainerOptions(pod *v1.Pod, container *v1.Container) *DeviceRunContainerOptions {
+	h.Lock()
+	defer h.Unlock()
+	return h.podDevices.deviceRunContainerOptions(string(pod.UID), container.Name)
+}
+
+// updateAllocatedDevices gets a list of active pods and then frees any Devices that are bound to
 // terminated pods. Returns error on failure.
-func (h *DevicePluginHandlerImpl) updateAllocatedDevices(activePods []*v1.Pod) {
+func (h *HandlerImpl) updateAllocatedDevices(activePods []*v1.Pod) {
 	h.Lock()
 	defer h.Unlock()
 	activePodUids := sets.NewString()
 	for _, pod := range activePods {
 		activePodUids.Insert(string(pod.UID))
 	}
-	for _, podDevs := range h.allocatedDevices {
-		allocatedPodUids := podDevs.pods()
-		podsToBeRemoved := allocatedPodUids.Difference(activePodUids)
-		glog.V(5).Infof("pods to be removed: %v", podsToBeRemoved.List())
-		podDevs.delete(podsToBeRemoved.List())
+	allocatedPodUids := h.podDevices.pods()
+	podsToBeRemoved := allocatedPodUids.Difference(activePodUids)
+	if len(podsToBeRemoved) <= 0 {
+		return
 	}
-}
-
-type checkpointEntry struct {
-	PodUID        string
-	ContainerName string
-	ResourceName  string
-	DeviceID      string
-}
-
-// checkpointData struct is used to store pod to device allocation information
-// in a checkpoint file.
-// TODO: add version control when we need to change checkpoint format.
-type checkpointData struct {
-	Entries []checkpointEntry
+	glog.V(5).Infof("pods to be removed: %v", podsToBeRemoved.List())
+	h.podDevices.delete(podsToBeRemoved.List())
+	// Regenerated allocatedDevices after we update pod allocation information.
+	h.allocatedDevices = h.podDevices.devices()
 }
 
 // Checkpoints device to container allocation information to disk.
-func (h *DevicePluginHandlerImpl) writeCheckpoint() error {
+func (h *HandlerImpl) writeCheckpoint() error {
+	h.Lock()
+	defer h.Unlock()
 	filepath := h.devicePluginManager.CheckpointFile()
-	var data checkpointData
-	for resourceName, podDev := range h.allocatedDevices {
-		for podUID, conDev := range podDev {
-			for conName, devs := range conDev {
-				for _, devId := range devs.UnsortedList() {
-					data.Entries = append(data.Entries, checkpointEntry{podUID, conName, resourceName, devId})
-				}
-			}
-		}
-	}
-	dataJson, err := json.Marshal(data)
+	data := h.podDevices.toCheckpointData()
+	dataJSON, err := json.Marshal(data)
 	if err != nil {
 		return err
 	}
-	return ioutil.WriteFile(filepath, dataJson, 0644)
+	return ioutil.WriteFile(filepath, dataJSON, 0644)
```

Comment:
This needs to be out of the critical section. Performance is going to be very impacted